### ISSUE TITLE
Evaluate forms in REPL namespace ignored in new socket repls

### ISSUE THREAD

**Issue by vlaaad** [cursive-ide/cursive#2244]
In 1.9.0-eap2, when using socket repl, it sends forms like these:
```clj
^{:clojure.core/eval-file "/some/file.clj" :line 1 :column 1} 
(+ 1 2 3)
```
This does not change `*ns*` to namespace of `file.clj`, like it previously did with also sending `(in-ns 'file-ns)`
Maybe Cursive should send `(in-ns ...)` as well?

**Comment by cursive-ide** [comment]
Just to be clear, you're talking about using the new socket REPL Remote REPL option, correct? That will send the forms as you describe, in Clojure 1.10+ that ensures that the form is compiled with the correct line/column numbers for stack traces and debugging. But under the hood the socket REPL isn't totally streaming, it actually sends structured forms including the namespace, and then the server side re-binds `*ns*` during eval. 

Are you seeing a change in behaviour?

**Comment by vlaaad** [comment]
Hmm, no, I use `clojure.main` with `-m my-custom-main`  which acts as REPL, so it's a local REPL.

**Comment by vlaaad** [comment]
Just starting `clj -J-Dclojure.server.repl='{:port 55555 :accept clojure.main/repl}'` and then connecting to it with Remote Socket REPL configuration does nothing, it's stuck on `Connecting to remote Socket REPL...`, what am I doing wrong?

**Comment by cursive-ide** [comment]
BTW for the rationale for this change, see #2113. As described in [this comment](https://github.com/cursive-ide/cursive/issues/2113#issuecomment-511249499), the previous behaviour is controlled by a flag which you can set in the run config. Cursive will warn you if it thinks you're using the wrong option, but you can just ignore that.

**Comment by cursive-ide** [comment]
I can't reproduce what you're seeing with the new socket REPL, this is what I see:
![Screen Shot 2019-10-09 at 13 39 56](https://user-images.githubusercontent.com/5589171/66443142-77738c80-ea2d-11e9-9365-a59081be176c.png)
This was using the following run config:
![Screen Shot 2019-10-09 at 13 40 16](https://user-images.githubusercontent.com/5589171/66443162-85c1a880-ea2d-11e9-9b3e-c885ad1f2707.png)
Are you seeing this hanging with a similar config?

**Comment by vlaaad** [comment]
I fully support #2113 change, just wanted to point out that behavior is different because it does not add `(in-ns ...)`, which previous did

**Comment by vlaaad** [comment]
There is a setting in settings dialog, named "Evaluate form in REPL namespace", with this change it's ignored.

**Comment by cursive-ide** [comment]
Yes, that's a fair comment. I'll think about this - previously the `(in-ns ...)` was possible because Cursive created a file and loaded it with `load-file`, which was the only way to get the correct file/line number. Since without the workaround it now just sends a form instead, I'll probably have to bind `*ns*` instead - not quite the same, but should achieve the same effect.

With this change, for clojure.main REPLs, that setting will be ignored for clojure.main REPLs until I implement the above, yes. It should still work for nREPL and the new socket REPLs though. 

Just to clarify - are you still finding that the new socket REPL is hanging at startup?

**Comment by vlaaad** [comment]
Alternatively you can send 2 forms: `(in-ns ...)` first, what actually sent second.

I can't reproduce socket repl hanging at startup :man_shrugging:

**Comment by cursive-ide** [comment]
The problem with sending two forms is that it permanently changes `*ns*` (i.e. the current namespace) for that REPL. What Cursive did previously only changed it temporarily for the single command that was sent - binding `*ns*` for each individual evaluation would also do that.

**Comment by vlaaad** [comment]
It might not work:
```clj
(ns a
  (:require [clojure.main :as m]))

(in-ns 'a)

::m/a
;; => :clojure.main/a

(ns b)

(in-ns 'b)

(binding [*ns* (the-ns 'a)]
  ::m/a)
; Syntax error reading source at (REPL:1322:8).
; Invalid token: ::m/a
```

**Comment by vlaaad** [comment]
`read` should happen in desired namespace.

**Comment by cursive-ide** [comment]
Yes, you're right. What I think that means is that for clojure.main REPLs, if the user wants the namespace switching feature (used when sending forms from an editor) then I'll have to use the `load-file` trick.

This particular failure is because `*ns*` isn't bound at read time, but there are other problems too. IIRC `*ns*` must be bound before `eval` is called. This works, but obscures the form actually executed for e.g. REBL as well:
```
(binding [*ns* (the-ns 'a)]
  (eval (read-string "::m/a")))
```
